providers/default/ciphers/cipher_chacha20_poly1305.c

```diff
@@ -0,0 +1,692 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Dispatch functions for chacha20_poly1305 cipher */
+
+#include "include/crypto/poly1305.h"
+#include "cipher_chacha20.h"
+#include "internal/provider_algs.h"
+#include "internal/providercommonerr.h"
+
+typedef struct {
+    POLY1305 poly1305;
+    PROV_CHACHA20_CTX chacha;
+    unsigned int nonce[12 / 4];
+    unsigned char tag[POLY1305_BLOCK_SIZE];
+    unsigned char tls_aad[POLY1305_BLOCK_SIZE];
+    struct { uint64_t aad, text; } len;
+    unsigned int enc : 1;
+    unsigned int aad : 1;
+    unsigned int iv_set : 1;
+    unsigned int mac_inited : 1;
+    size_t tag_len, nonce_len;
+    size_t tls_payload_length;
+    size_t tls_aad_pad_sz;
+} PROV_CHACHA_AEAD_CTX;
+
+#define NO_TLS_PAYLOAD_LENGTH ((size_t)-1)
+#define CHACHA20_POLY1305_KEYLEN CHACHA_KEY_SIZE
+#define CHACHA20_POLY1305_BLKLEN 1
+#define CHACHA20_POLY1305_IVLEN 12
+#define CHACHA20_POLY1305_MAX_IVLEN 12
+#define CHACHA20_POLY1305_MODE 0
+#define CHACHA20_POLY1305_FLAGS (EVP_CIPH_FLAG_AEAD_CIPHER                     \
+                                | EVP_CIPH_ALWAYS_CALL_INIT                    \
+                                | EVP_CIPH_CTRL_INIT                           \
+                                | EVP_CIPH_CUSTOM_COPY                         \
+                                | EVP_CIPH_FLAG_CUSTOM_CIPHER                  \
```

Comment:
Mostly only used in the legacy path - although there seems to be some very strange usage of this in libssl.